pkg/hashreader/hashreader.go

```diff
@@ -0,0 +1,120 @@
+package hashreader
+
+// HashReader writes what it reads from an io.Reader to an MD5 and SHA256 hash.Hash.
+import (
+	"bytes"
+	"crypto/md5"
+	"crypto/sha256"
+	"encoding/hex"
+	"hash"
+	"io"
+)
+
+// HashReader verifies that the content of the io.Reader matches the expected checksums.
+type HashReader struct {
+	md5sum, sha256sum []byte // Byte values of md5sum, sha256sum of client sent values.
+
+	src                 io.Reader
+	size                int64
+	md5Hash, sha256Hash hash.Hash
+}
+
+// New returns a new HashReader computing the MD5 sum and SHA256 sum
+// (if set) of the provided io.Reader.
+func New(src io.Reader, size int64, md5Hex, sha256Hex string) (*HashReader, error) {
+	sha256sum, err := hex.DecodeString(sha256Hex)
+	if err != nil {
+		return nil, err
+	}
+	md5sum, err := hex.DecodeString(md5Hex)
+	if err != nil {
+		return nil, err
+	}
+	if hr, ok := src.(*HashReader); ok {
+		return hr, nil
```

Comment:
As we found given md5/sha256 are valid and if `src` is an `HashReader` I guess we need to error out about nesting is not allowed.  This way we can avoid confusion i.e. user might think newly provided checksums are used in the returned `HashReader`

Reply:
Yes good idea.